### PR TITLE
fix: bpapi badfun again

### PR DESCRIPTION
**Pull request by zhongwencool**
Fixes <issue-or-jira-number>

Release version: v/e5.?

## PR Checklist
Please convert it to a draft if any of the following conditions are not met. Reviewers may skip over until all the items are checked:

- [ ] Added tests for the changes
- [ ] Added property-based tests for code which performs user input validation
- [ ] Changed lines covered in coverage report
- [ ] Change log has been added to `changes/(ce|ee)/(feat|perf|fix|breaking)-<PR-id>.en.md` files
- [ ] For internal contributor: there is a jira ticket to track this change
- [ ] Created PR to [emqx-docs](https://github.com/emqx/emqx-docs) if documentation update is required, or link to a follow-up jira ticket
- [ ] Schema changes are backward compatible

## Checklist for CI (.github/workflows) changes

- [ ] If changed package build workflow, pass [this action](https://github.com/emqx/emqx/actions/workflows/build_packages.yaml) (manual trigger)
- [ ] Change log has been added to `changes/` dir for user-facing artifacts update

### CODE REVIEW

**Review by HJianBo — APPROVED** [review]

**Comment by id** [comment]
```
apps/emqx/src/bpapi/emqx_bpapi.erl
Line 114 Column 29: Guard test [{#bpapi{key::{'$1','$2'},version::'$3'},[{'=/=','$1','multicall'} | {'=:=','$2',{'const',_}},...],['$3',...]},...] =:= MS::[{#bpapi{key::{'$1','$2'},version::'_'},[{'=:=','$1',{'const',_}},...],[{{'$1','$2'}},...]},...] can never succeed
```

File: apps/emqx/src/bpapi/emqx_bpapi.erl [hunk 1]
```diff
@@ -109,7 +109,33 @@ announce(Node, App) ->
                      || {API, Version} <- Data
                     ],
                     %% Update maximum supported version:
-                    _ = [update_minimum(API) || {API, _} <- Data],
+                    _ = [
```
[the remark this answers is not in the document]
Comment (zmstone) [reply apps/emqx/src/bpapi/emqx_bpapi.erl:112]: the function has to be remote call.
or an *exact* local call reference which has to be ordered the same among all nodes
e.g.
func1 being the Nth defined function in this module in all nodes.

File: apps/emqx/src/bpapi/emqx_bpapi.erl [hunk 2]
```diff
@@ -109,7 +109,33 @@ announce(Node, App) ->
...
+                        begin
+                            %%Line 114 Column 29: Guard test [{#bpapi{key::{'$1','$2'},version::'$3'},
+                            %% [{'=/=','$1','multicall'}]},...] can never succeed
+                            %%MS = ets:fun2ms(fun(
```
Comment (zmstone) [inline apps/emqx/src/bpapi/emqx_bpapi.erl:116]: or: move this part out of the closure

**Comment by zhongwencool** [comment]
I call those function from replicant's shell is ok.
```
{atomic, ok} = mria:transaction(
                emqx_common_shard,
                fun() ->
                    MS = ets:fun2ms(fun(#bpapi{key = {N, API}}) when N =:= Node ->
                        {N, API}
                    end),
                    OldKeys = mnesia:select(bpapi, MS, write),
                    _ = [
                        mnesia:delete({bpapi, Key})
                     || Key <- OldKeys
                    ],
                    %% Insert new records:
                    _ = [
                        mnesia:write(#bpapi{key = {Node, API}, version = Version})
                     || {API, Version} <- Data
                    ],
                    %% Update maximum supported version:
                    lists:foreach(
                        fun({API, _}) ->
                            MS = [
                                {
                                    #bpapi{key = {'$1', '$2'}, version = '$3'},
                                    [{'=/=', '$1', multicall}, {'=:=', '$2', {const, API}}],
                                    ['$3']
                                }
                            ],
                            MinVersion = lists:min(mnesia:select(bpapi, MS)),
                            mnesia:write(#bpapi{key = {multicall, API}, version = MinVersion})
                        end,
                        Data
                    )
                end
            ),
            ok
```
But it's also failed will the badfun when replicant start sometime (not all the times)...

```
2024-03-26T08:21:33.936776+00:00 [error] crasher: initial call: application_master:init/4, pid: <0.3473.0>, registered_name: [], exit: {{bad_return,{{emqx_app,start,[normal,[]]},{'EXIT',{{badmatch,{aborted,{{badfun,#Fun<emqx_bpapi.0.117318672>},[{mria_upstream,'-transactional_wrapper/3-fun-0-',4,[{file,"mria_upstream.erl"},{line,52}]},{mria_upstream,'-transactional_wrapper/3-fun-0-',4,[{file,"mria_upstream.erl"},{line,52}]},{mnesia_tm,apply_fun,3,[{file,"mnesia_tm.erl"},{line,889}]},{mnesia_tm,execute_transaction,5,[{file,"mnesia_tm.erl"},{line,865}]},{erpc,execute_call,4,[{file,"erpc.erl"},{line,589}]}]}}},[{emqx_bpapi,announce,2,[{file,"emqx_bpapi.erl"},{line,95}]},{emqx_app,start,2,[{file,"emqx_app.erl"},{line,43}]},{application_master,start_it_old,4,[{file,"application_master.erl"},{line,293}]}]}}}},[{application_master,init,4,[{file,"application_master.erl"},{line,142}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,241}]}]}, ancestors: [<0.3472.0>], message_queue_len: 1, messages: 
```

**Comment by ieQu1** [comment]
This should prevent badfun's: https://github.com/emqx/emqx/pull/12786

**Comment by zhongwencool** [comment]
close because https://github.com/emqx/emqx/pull/12786

**Pull request closed without merging**